#### Data Analysis with pandas

**Use-case Scenarion:**

You are provided with data (test_data.xlsx) from a retail shop that consists of 8 columns(Invoice, StockCode, Description, Quantity, InvoiceDate, Price, Customer ID, and Country).
You are supposed to analyze data in indexes between (433200, 437200)! Do not forget to clean the data of NaNs, as failing to do so may render your answers invalid!

*Fill in your answers to the text below, as if it were your report from the data.*



1.   The amount of **unique invoices** in the provided dataset for indexes (433200, 437200) is equal to --?-- (integer).

2.   The **country with the most orders** is --?-- (string) with the **amount** of --?-- (integer).

3.   The **country with the least orders** is --?-- (string) with the amount of --?-- (integer).
4.   The **highest total price per item** in one order was --?--  (float) in order with **id** --?-- (integer).
5.   The **second highest spending customer** with id --?-- The second highest spending customer with id --?--(integer) **times**.


**Hints:**
Total price per item is calculated as price times quantity.
Methods to consider: *dropna, group_by, sort_values*

In [1]:
# Loading Lib
import pandas as pd

In [2]:
# Loading Data Note: Important to spacify the sheet you are working on
data1 = pd.read_excel('/content/test_data.xlsx', sheet_name = 'Year 2009-2010')

In [3]:
# slicing our desired Data
data1 = data1.iloc[433200: 437200]
data1

Invoice StockCode                          Description  Quantity  \
433200  530811    16206B            RED PURSE WITH PINK HEART         3   
433201  530811     22662          LUNCH BAG DOLLY GIRL DESIGN         1   
433202  530811     20727              LUNCH BAG  BLACK SKULL.         2   
433203  530811     20726                   LUNCH BAG WOODLAND         3   
433204  530811     22382           LUNCH BAG SPACEBOY DESIGN          2   
...        ...       ...                                  ...       ...   
437195  531210    85071B   RED CHARLIE+LOLA PERSONAL DOORSIGN         1   
437196  531210    85071A  BLUE CHARLIE+LOLA PERSONAL DOORSIGN         1   
437197  531210     22898        CHILDRENS APRON APPLES DESIGN         1   
437198  531210     22899         CHILDREN'S APRON DOLLY GIRL          1   
437199  531210     21788                   KIDS RAIN MAC BLUE         2   

               InvoiceDate  Price  Customer ID         Country  
433200 2010-11-04 13:04:00   1.25      13137.0  United Kingdom  
433201 2010-11-04 13:04:00   1.65      13137.0  United Kingdom  
433202 2010-11-04 13:04:00   1.65      13137.0  United Kingdom  
433203 2010-11-04 13:04:00   1.65      13137.0  United Kingdom  
433204 2010-11-04 13:04:00   1.65      13137.0  United Kingdom  
...                    ...    ...          ...             ...  
437195 2010-11-05 16:07:00   2.95      13263.0  United Kingdom  
437196 2010-11-05 16:07:00   2.95      13263.0  United Kingdom  
437197 2010-11-05 16:07:00   1.95      13263.0  United Kingdom  
437198 2010-11-05 16:07:00   2.10      13263.0  United Kingdom  
437199 2010-11-05 16:07:00   0.85      13263.0  United Kingdom  

[3028 rows x 8 columns]

In [192]:
# we must have information about the data that we are working with
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 433200 to 437199
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Invoice      4000 non-null   object        
 1   StockCode    4000 non-null   object        
 2   Description  3917 non-null   object        
 3   Quantity     4000 non-null   int64         
 4   InvoiceDate  4000 non-null   datetime64[ns]
 5   Price        4000 non-null   float64       
 6   Customer ID  3028 non-null   float64       
 7   Country      4000 non-null   object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 250.1+ KB


In [189]:
data1.describe()

Quantity        Price   Customer ID
count  4000.000000  4000.000000   3028.000000
mean      5.012500     3.741615  15154.674373
std     177.398748     8.183365   1865.822694
min   -8800.000000     0.000000  12393.000000
25%       1.000000     1.250000  13615.000000
50%       3.000000     2.510000  15023.000000
75%      10.000000     4.250000  16919.000000
max    1152.000000   341.280000  18259.000000

In [ ]:
# Droping Null Values in axis=0 means droping whole Row which is containg it
df = data1.dropna(axis=0)
df

In [194]:
# Question 1.
len(df.Invoice.unique())

171

In [5]:
# max(df.Country)

'United Kingdom'

In [8]:
# min(df.Country)

'Australia'

In [ ]:
# created new col with Total sale of one item
df['Total'] = df.Quantity * df.Price

In [232]:
# Question 2
max_cn = df[(df.Country == max(df.Country))]['Total'].count()
print(f"Country {max(df.Country)} with maximum orders of : {max_cn}")

Country United Kingdom with maximum orders of : 2688


In [233]:
# Question 3
min_cn = df[(df.Country == min(df.Country))]['Total'].count()
print(f"Country {min(df.Country)} with minimum orders of : {min_cn}")

Country Australia with minimum orders of : 5


In [176]:
# single item with highest Total price
# df[df.Total == df['Total'].max()]
# Highest ordered Quantity
# df[df.Quantity == df['Quantity'].max()]

Invoice StockCode                         Description  Quantity  \
436520  531069     22734  SET OF 6 RIBBONS VINTAGE CHRISTMAS      1152   

               InvoiceDate  Price  Customer ID         Country   Total  
436520 2010-11-05 12:10:00    1.8      17949.0  United Kingdom  2073.6

In [201]:
# Question 4.
df.groupby(df.Invoice)['Total'].sum().nlargest(1)#.sort_values(ascending=False)

Invoice
531204    5321.16
Name: Total, dtype: float64

In [231]:
# Question 5
# find Id of the
orders_sum = df.groupby(df['Customer ID'])['Total'].sum().sort_values(ascending=False).nlargest(3)
print("ids wiht aorders amount\n", orders_sum)
norders = df[df['Customer ID'] == 12477.0].Invoice.nunique()
print("\nNo of the Times the seconed Highest spending customer ordered = ", norders)

ids wiht aorders amount
 Customer ID
17389.0    6412.14
12477.0    2570.77
17949.0    2496.96
Name: Total, dtype: float64

No of the Times the seconed Highest spending customer ordered =  1


#Report:

1.   The amount of **unique invoices** in the provided dataset for indexes (433200, 437200) is equal to 171 (integer).

2.   The **country with the most orders** is ***United Kingdom*** (string) with the **amount** of 2688 (integer).

3.   The **country with the least orders** is ***Australia*** (string) with the **amount** of 5 (integer).
4.   The **highest total price per item** in one order was 5321.16  (float) in order with **id** 531204 (integer).
5.   The **second highest spending customer** with id 12477 The second highest spending customer with id 1 (integer) **times**.